conn.go

```diff
@@ -47,49 +57,49 @@ func NewConn(gw string, crt string, key string) (Conn, error) {
 func NewConnWithFiles(gw string, certFile string, keyFile string) (Conn, error) {
 	cert, err := tls.LoadX509KeyPair(certFile, keyFile)
 	if err != nil {
-		return Conn{}, err
+		return &conn{}, err
 	}
 
 	return NewConnWithCert(gw, cert), nil
 }
 
 // Connect actually creates the TLS connection
-func (c *Conn) Connect() error {
+func (c *conn) Connect() error {
 	// Make sure the existing connection is closed
-	if c.NetConn != nil {
-		c.NetConn.Close()
+	if c.netConn != nil {
+		c.netConn.Close()
 	}
 
-	conn, err := net.Dial("tcp", c.gateway)
+	tlsConn, err := tls.Dial("tcp", c.gateway, c.tls)
 	if err != nil {
 		return err
 	}
 
-	tlsConn := tls.Client(conn, c.Conf)
-	err = tlsConn.Handshake()
-	if err != nil {
-		return err
-	}
-
-	c.NetConn = tlsConn
+	c.netConn = tlsConn
 	return nil
 }
 
-func (c *Conn) Close() error {
-	if c.NetConn != nil {
-		return c.NetConn.Close()
+func (c *conn) Close() error {
+	if c.netConn != nil {
+		return c.netConn.Close()
 	}
 
 	return nil
 }
 
 // Read reads data from the connection
-func (c *Conn) Read(p []byte) (int, error) {
-	i, err := c.NetConn.Read(p)
-	return i, err
+func (c *conn) Read(p []byte) (int, error) {
+	return c.netConn.Read(p)
+}
+
+// ReadWithTimeout reads data from the connection and returns an error
+// after duration
+func (c *conn) ReadWithTimeout(p []byte, deadline time.Time) (int, error) {
+	c.netConn.SetReadDeadline(deadline)
```

Comment:
Tracing this down to setDeadlineImpl, it seems to do a bunch of work. Do we want to do that on every read? Having a separate SetReadDeadline method seems more flexible.